Code/Framework/AzCore/AzCore/DOM/DomPatch.h

```diff
@@ -0,0 +1,206 @@
+        //! Patching strategy that attemps to apply all operations in a patch, but ignores operation failures and continues.
+        void IgnoreFailureAndContinue(PatchApplicationState& state);
+    } // namespace PatchApplicationStrategy
+
+    //! A set of operations that can be applied to a Value to produce a new Value.
+    //! \see PatchOperation
+    class Patch final
+    {
+    public:
+        using StrategyFunctor = AZStd::function<void(PatchApplicationState&)>;
+        using OperationsContainer = AZStd::vector<PatchOperation>;
+
+        Patch() = default;
+        Patch(const Patch&) = default;
+        Patch(Patch&&) = default;
+        Patch(AZStd::initializer_list<PatchOperation> init);
+
+        template<class InputIterator>
+        Patch(InputIterator first, InputIterator last)
+            : m_operations(first, last)
+        {
+        }
+
+        Patch& operator=(const Patch&) = default;
+        Patch& operator=(Patch&&) = default;
+
+        bool operator==(const Patch& rhs) const;
+        bool operator!=(const Patch& rhs) const;
+
+        const OperationsContainer& GetOperations() const;
+        void PushBack(PatchOperation op);
+        void PushFront(PatchOperation op);
+        void Pop();
+        void Clear();
+        const PatchOperation& At(size_t index) const;
+        size_t Size() const;
+
+        PatchOperation& operator[](size_t index);
+        const PatchOperation& operator[](size_t index) const;
+
+        OperationsContainer::iterator begin();
+        OperationsContainer::iterator end();
+        OperationsContainer::const_iterator begin() const;
+        OperationsContainer::const_iterator end() const;
+        OperationsContainer::const_iterator cbegin() const;
+        OperationsContainer::const_iterator cend() const;
+        size_t size() const;
+
+        AZ::Outcome<Value, AZStd::string> Apply(Value rootElement, StrategyFunctor strategy = PatchApplicationStrategy::HaltOnFailure) const;
+        AZ::Outcome<void, AZStd::string> ApplyInPlace(Value& rootElement, StrategyFunctor strategy = PatchApplicationStrategy::HaltOnFailure) const;
+
+        Value GetDomRepresentation() const;
+        static AZ::Outcome<Patch, AZStd::string> CreateFromDomRepresentation(Value domValue);
+
+    private:
+        OperationsContainer m_operations;
+    };
+
+    //! A set of patches for applying a change and doing the inverse operation (i.e. undoing it).
+    struct PatchInfo
+    {
+        Patch m_forwardPatches;
+        Patch m_inversePatches;
+    };
+
+    //! Parameters for GenerateHierarchicalDeltaPatch.
+    struct DeltaPatchGenerationParameters
+    {
+        static constexpr size_t NoReplace = AZStd::numeric_limits<size_t>::max();
+
+        //! The threshold of changed values in a node or array which, if exceeded, will cause the generation to create an
+        //! entire "replace" oepration instead. If set to NoReplace, no replacement will occur.
+        size_t m_replaceThreshold = 3;
```

Comment:
Should there also be a default value to allow the collapse of "replace"s to never happen? Something like
`static constexpr size_t AlwaysFullReplace = AZStd::numeric_limits<size_t>::max()-1;`

Reply:
Added an `AlwaysFullReplace` set to `0` that should provide this behavior.